src/loader/init.c

```diff
@@ -0,0 +1,149 @@
+#include <postgres.h>
+#include <pg_config.h>
+#include <access/xact.h>
+#include <commands/extension.h>
+#include <miscadmin.h>
+#include <utils/guc.h>
+#include <utils/inval.h>
+#include <parser/analyze.h>
+#include <nodes/print.h>
+
+
+#include "extension.h"
+
+/*
+ * Some notes on design:
+ *
+ * We do not check for the installation of the extension upon loading the extension and instead rely on a hook for two reasons:
+ * 1) We probably can't
+ *	- The shared_preload_libraries is called in PostmasterMain which is way before InitPostgres is called.
+ *			(Note: This happens even before the fork of the backend)
+ *	-- This means we cannot query for the existance of the extension yet because the caches are initialized in InitPostgres.
+ * 2) We actually don't want to load the extension in two cases:
+ *	  a) We are upgrading the extension.
+ *	  b) We set the guc timescaledb.disable_load.
+ *
+ *
+ */
+
+#define MIN_SUPPORTED_VERSION_STR "9.6"
+#define MIN_SUPPORTED_VERSION_NUM 90600
+
+#if PG_VERSION_NUM < MIN_SUPPORTED_VERSION_NUM
+#error "Unsupported version of PostgreSQL. Check src/init.c for supported versions."
+#endif
+
+#ifdef PG_MODULE_MAGIC
+PG_MODULE_MAGIC;
+#endif
+
+#define GUC_DISABLE_LOAD_NAME "timescaledb.disable_load"
+
+extern void _PG_init(void);
+extern void _PG_fini(void);
+
+bool		guc_disable_load = false;
+
+static void
+inval_cache_callback(Datum arg, Oid relid)
+{
+	if (guc_disable_load)
+		return;
+	extension_check();
+}
+
+static void
+post_analyze_hook(ParseState *pstate, Query *query)
+{
+	if (guc_disable_load)
+		return;
+
+	/* Don't do a load if setting timescaledb.disable_load or doing an update */
+	if (query->commandType == CMD_UTILITY)
+	{
+		if (IsA(query->utilityStmt, VariableSetStmt))
+		{
+			VariableSetStmt *stmt = (VariableSetStmt *) query->utilityStmt;
+
+			if (strcmp(stmt->name, GUC_DISABLE_LOAD_NAME) == 0)
+			{
+				return;
+			}
+		}
+		else if (IsA(query->utilityStmt, AlterExtensionStmt))
+		{
+			AlterExtensionStmt *stmt = (AlterExtensionStmt *) query->utilityStmt;
+
+			if (strcmp(stmt->extname, EXTENSION_NAME) == 0)
+			{
+				if (extension_loaded())
+				{
+					ereport(ERROR,
+							(errmsg("Cannot update the extension after the old version has already been loaded"),
+							 errhint("You should start a new session and execute ALTER EXTENSION as the first command")));
+				}
+
+				return;
+			}
+		}
+	}
+	extension_check();
+}
+
+void
+_PG_init(void)
+{
+	if (!process_shared_preload_libraries_in_progress)
+	{
+		/* cannot use GUC variable here since extension not yet loaded */
+		char	   *allow_install_without_preload = GetConfigOptionByName("timescaledb.allow_install_without_preload", NULL, true);
+
+		if (allow_install_without_preload == NULL ||
+			strlen(allow_install_without_preload) != 2 ||
```

Comment:
This could be replaced with a single `strcmp(allow_install_without_preload, "on") != 0`, with the same result.